dynamic_instrumentation/src/main.rs

```diff
@@ -0,0 +1,413 @@
+
+    // TODO(kkysen) We probably have to delete binaries that have different names from the crates.
+
+    // Delete all executables in `target/${profile}/deps/` starting with the crate name + `-`.
+    let target_read_dir = match cargo_metadata.target_directory.read_dir() {
+        Ok(read_dir) => Ok(read_dir),
+        Err(e) => if e.kind() == ErrorKind::NotFound {
+            return Ok(());
+        } else {
+            Err(e)
+        }
+    }?;
+    for profile_dir in target_read_dir {
+        let profile_dir = profile_dir?;
+        if !profile_dir.file_type()?.is_dir() {
+            continue;
+        }
+        for artifact in profile_dir.path().join("deps").read_dir()? {
+            let artifact = artifact?;
+            if !artifact.file_type()?.is_file() {
+                continue;
+            }
+            let file_name = artifact.file_name();
+            let file_name = file_name.to_str();
+            for crate_target in &info.crate_targets {
+                let prefix = format!("{}-", crate_target.crate_name.as_str());
+                // [`Path::starts_with`] only checks whole components at once,
+                // and `OsStr::starts_with` doesn't exist yet.
+                if file_name
+                    .map(|name| name.starts_with(&prefix))
+                    .unwrap_or_default()
+                {
+                    let artifact = artifact.path();
+                    if artifact.is_executable() {
+                        fs_err::remove_file(&artifact)?;
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn main() -> eyre::Result<()> {
+    color_eyre::install()?;
+    env_logger::init();
+
+    let rustc_wrapper_var = "RUSTC_WRAPPER";
+    let instrument_info_var = "C2RUST_INSTRUMENT_INFO";
+
+    let own_exe = env::current_exe()?;
+
+    let wrapping_rustc = env::var_os(rustc_wrapper_var).as_deref() == Some(own_exe.as_os_str());
+    if wrapping_rustc {
+        let sysroot = get_sysroot()?;
+        let at_args = env::args().skip(1).collect::<Vec<_>>();
+        let crate_target = CrateTarget::from_rustc_args(&at_args)?;
+        let info = env::var(instrument_info_var)?;
+        let info = serde_json::from_str::<InstrumentInfo>(&info)?;
+        let should_instrument = info.crate_targets.contains(&crate_target);
+        if should_instrument {
+            instrument_rustc(at_args, &sysroot, &info.metadata)?;
+        } else {
+            let status = passthrough_rustc(&at_args, &sysroot)?;
+            exit_with_status(status);
+        }
+    } else {
+        let Args {
+            metadata,
+            cargo_args,
+        } = Args::parse();
+
+        let cargo_metadata = MetadataCommand::new().exec()?;
+        let crate_targets = CrateTarget::from_metadata(&cargo_metadata)?;
+        let info = InstrumentInfo {
+            crate_targets,
+            metadata,
+        };
+
+        delete_metadata_and_dependencies(&info, &cargo_metadata)?;
```

Comment:
Instead of deleting files ourselves, could you use `cargo clean --package crate_name` instead?

Reply:
Ah, I didn't know that's a thing.  In that case, that should work perfectly.  Thanks!

Reply:
Fixed in https://github.com/immunant/c2rust/pull/554/commits/035670ec7b8eb95592ce9f28aed720956e583b8d.  That's a lot cleaner and reliable, too!